### PR TITLE
Update factory query to use fts for addresses

### PR DESCRIPTION
**Pull request by trevormcdonald**
## Proposed change/fix

Add address querying support with postgres full text search, see the [database crate](https://github.com/target/consensource-database/pull/11) for more info.

## Types of changes

What types of changes does this pull request introduce to ConsenSource? _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (could be a small readme update, documentation contribution, etc.)

## How to run/test

1. Clone down the [consensource-compose](https://github.com/target/consensource-compose) repo:
```bash
 git clone --recurse-submodules https://github.com/target/consensource-compose.git
```
2. Make sure each submodule is up-to-date with the latest code changes
```sh
 git submodule update --remote $api
```
3. `cd api` and checkout this branch for api submodule
4. Build a local image for this image
```sh
 ./docker-helper.sh --build api
```
5. Build local images for other submodules as well if needed (`./docker-helper.sh --build local`)
6. Run `docker-compose -f docker-compose.run-local.yaml up`
7. Exec into the CLI container
8. Create an agent
```bash
 csrc agent create cliagent --url http://api:9009
```
9. Create an ingestion organization
```bash
 csrc organization create asserterorg 4 assertercontactname 7632918628 ENG --city minneapolis --country usa --street_address '821 thor street' --url http://api:9009
```
10. Create a factory assertion
```
 csrc assertion factory create {org_id_here} testfactoryassertion1 yuki 6123982765 JPN --city osaka --country japan --street_address 'fukuoka south st' --state_province kanto --postal_code 55414 --url http://api:9009
```
```
csrc assertion factory create {org_id_here} testfactoryassertion2 ren 6123982765 CHN --city shenzhen --country china --street_address 'dong east st' --state_province guangdong --postal_code 55446 --url http://api:9009
```
11. Query the API with one of the newly implemented parameters.
```
 http://localhost:9009/api/factories?address=shenzhen
```

### CODE REVIEW

File: src/route_handlers/factories.rs [hunk 1]
```diff
@@ -168,6 +172,21 @@ fn query_factories(
         count_query = count_query.filter(organizations::name.eq(name));
     }
 
+    if let Some(address) = params.address {
+        let org_ids: Vec<String> = addresses::table
+            .select(ADDRESS_COLUMNS)
+            .filter(addresses::start_block_num.le(head_block_num))
```
Comment (Patrick-Erichsen) [inline src/route_handlers/factories.rs:178]: Is there ever a case where we don't end up filtering with start/end block equal to the head block num? What do you think about doing a bit of cleanup work here to make a `filter_active_blocks(head_block_num)` method or something similar? 

Wouldn't need to update any other methods, but we could start doing it piecemeal as we update more code.

File: src/route_handlers/factories.rs [hunk 2]
```diff
@@ -168,6 +172,21 @@ fn query_factories(
         count_query = count_query.filter(organizations::name.eq(name));
     }
 
+    if let Some(address) = params.address {
```
Comment (Patrick-Erichsen) [inline src/route_handlers/factories.rs:175]: I think a comment here explaining that `address` is used for full text search would be nice

**Review by Patrick-Erichsen — APPROVED** [review]
LGTM 👍

File: src/route_handlers/factories.rs (hunk 1 above)
Comment (trevormcdonald) [reply src/route_handlers/factories.rs:178]: Currently no, the api only returns data not marked as stale at that head block.

I would be more in favor of a query build method that would encompass even more than just block numbers. An example that includes an id filter could look like:

```
pub fn build_query(table: Table, id: String, head_block_num: i64) -> Filter {
    table
    .filter(table::start_block_num.le(head_block_num))
    .filter(table::end_block_num.gt(head_block_num))
    .filter(table::table_id.eq(id.to_string()))
}
```

That doesn't quite work but I think it shows what I'm trying to get at. The actual implementation would be better served by using a macro that can take into account column names better.

I think an even better approach would be to move most of this code to the database repo. That would bring it in line with how the sds saves to the db and would let the api code not have to worry about db stuff, like it wouldn't even need to depend on diesel.

Comment (Patrick-Erichsen) [reply src/route_handlers/factories.rs:178]: 👀  

I agree that a lot more boilerplate than just the start/end block filtering could be included in a more generalized query method. Seems like a fair amount of additional work compared to what I proposed though so probably a `backlog` item for now.

**Review by Patrick-Erichsen — APPROVED** [review]

**Pull request merged**